File: src/main/java/dev/orscathinus/cavegame/util/FixedStepTimer.java
package dev.orscathinus.cavegame.util;

/**
 * Converts variable frame times into a bounded number of fixed simulation updates.
 * Rendering can run once per frame regardless of the update count.
 */
public final class FixedStepTimer {
    private static final double EPSILON = 1.0e-12;

    private final double stepSeconds;
    private final double maxFrameDeltaSeconds;
    private final int maxUpdatesPerFrame;

    private double previousTimeSeconds;
    private double accumulatorSeconds;
    private boolean initialized;

    public FixedStepTimer(
            double updatesPerSecond,
            double maxFrameDeltaSeconds,
            int maxUpdatesPerFrame
    ) {
        if (!Double.isFinite(updatesPerSecond) || updatesPerSecond <= 0.0) {
            throw new IllegalArgumentException("updatesPerSecond must be finite and positive");
        }
        if (!Double.isFinite(maxFrameDeltaSeconds) || maxFrameDeltaSeconds <= 0.0) {
            throw new IllegalArgumentException("maxFrameDeltaSeconds must be finite and positive");
        }
        if (maxUpdatesPerFrame <= 0) {
            throw new IllegalArgumentException("maxUpdatesPerFrame must be positive");
        }

        this.stepSeconds = 1.0 / updatesPerSecond;
        this.maxFrameDeltaSeconds = maxFrameDeltaSeconds;
        this.maxUpdatesPerFrame = maxUpdatesPerFrame;
    }

    public void reset(double currentTimeSeconds) {
        requireFiniteTime(currentTimeSeconds);
        previousTimeSeconds = currentTimeSeconds;
        accumulatorSeconds = 0.0;
        initialized = true;
    }

    public Frame advance(double currentTimeSeconds) {
        requireFiniteTime(currentTimeSeconds);
        if (!initialized) {
            reset(currentTimeSeconds);
            return new Frame(0.0, 0.0, 0, 0.0, false);
        }

        double rawDeltaSeconds = Math.max(currentTimeSeconds - previousTimeSeconds, 0.0);
        previousTimeSeconds = currentTimeSeconds;

        double acceptedDeltaSeconds = Math.min(rawDeltaSeconds, maxFrameDeltaSeconds);
        accumulatorSeconds += acceptedDeltaSeconds;

        int updateCount = 0;
        while (accumulatorSeconds + EPSILON >= stepSeconds
                && updateCount < maxUpdatesPerFrame) {
            accumulatorSeconds -= stepSeconds;
            updateCount++;
        }

        boolean discardedBacklog = accumulatorSeconds + EPSILON >= stepSeconds;
        if (discardedBacklog) {
            accumulatorSeconds %= stepSeconds;
        }
        if (accumulatorSeconds < 0.0 && accumulatorSeconds > -EPSILON) {
            accumulatorSeconds = 0.0;
        }

        double interpolationAlpha = accumulatorSeconds / stepSeconds;
        return new Frame(
                rawDeltaSeconds,
                acceptedDeltaSeconds,
                updateCount,
                interpolationAlpha,
                discardedBacklog
        );
    }

    public double stepSeconds() {
        return stepSeconds;
    }

    private static void requireFiniteTime(double timeSeconds) {
        if (!Double.isFinite(timeSeconds)) {
            throw new IllegalArgumentException("timeSeconds must be finite");
        }
    }

    public record Frame(
            double rawDeltaSeconds,
            double acceptedDeltaSeconds,
            int updateCount,
            double interpolationAlpha,
            boolean discardedBacklog
    ) {
    }
}

File: web/fixed-step-timer.mjs
const EPSILON = 1.0e-12;

/**
 * Converts variable animation-frame timestamps into bounded fixed updates.
 * Rendering still occurs once per requestAnimationFrame callback.
 */
export class FixedStepTimer {
    #stepSeconds;
    #maxFrameDeltaSeconds;
    #maxUpdatesPerFrame;
    #previousTimeSeconds = 0;
    #accumulatorSeconds = 0;
    #initialized = false;

    constructor({
        updatesPerSecond = 60,
        maxFrameDeltaSeconds = 0.25,
        maxUpdatesPerFrame = 5,
    } = {}) {
        if (!Number.isFinite(updatesPerSecond) || updatesPerSecond <= 0) {
            throw new RangeError("updatesPerSecond must be finite and positive");
        }
        if (!Number.isFinite(maxFrameDeltaSeconds) || maxFrameDeltaSeconds <= 0) {
            throw new RangeError("maxFrameDeltaSeconds must be finite and positive");
        }
        if (!Number.isInteger(maxUpdatesPerFrame) || maxUpdatesPerFrame <= 0) {
            throw new RangeError("maxUpdatesPerFrame must be a positive integer");
        }

        this.#stepSeconds = 1 / updatesPerSecond;
        this.#maxFrameDeltaSeconds = maxFrameDeltaSeconds;
        this.#maxUpdatesPerFrame = maxUpdatesPerFrame;
    }

    get stepSeconds() {
        return this.#stepSeconds;
    }

    reset(timestampMilliseconds) {
        const currentTimeSeconds = toFiniteSeconds(timestampMilliseconds);
        this.#previousTimeSeconds = currentTimeSeconds;
        this.#accumulatorSeconds = 0;
        this.#initialized = true;
    }

    advance(timestampMilliseconds) {
        const currentTimeSeconds = toFiniteSeconds(timestampMilliseconds);
        if (!this.#initialized) {
            this.reset(timestampMilliseconds);
            return Object.freeze({
                rawDeltaSeconds: 0,
                acceptedDeltaSeconds: 0,
                updateCount: 0,
                interpolationAlpha: 0,
                discardedBacklog: false,
            });
        }

        const rawDeltaSeconds = Math.max(
            currentTimeSeconds - this.#previousTimeSeconds,
            0,
        );
        this.#previousTimeSeconds = currentTimeSeconds;

        const acceptedDeltaSeconds = Math.min(
            rawDeltaSeconds,
            this.#maxFrameDeltaSeconds,
        );
        this.#accumulatorSeconds += acceptedDeltaSeconds;

        let updateCount = 0;
        while (
            this.#accumulatorSeconds + EPSILON >= this.#stepSeconds
            && updateCount < this.#maxUpdatesPerFrame
        ) {
            this.#accumulatorSeconds -= this.#stepSeconds;
            updateCount += 1;
        }

        const discardedBacklog =
            this.#accumulatorSeconds + EPSILON >= this.#stepSeconds;
        if (discardedBacklog) {
            this.#accumulatorSeconds %= this.#stepSeconds;
        }
        if (this.#accumulatorSeconds < 0 && this.#accumulatorSeconds > -EPSILON) {
            this.#accumulatorSeconds = 0;
        }

        return Object.freeze({
            rawDeltaSeconds,
            acceptedDeltaSeconds,
            updateCount,
            interpolationAlpha: this.#accumulatorSeconds / this.#stepSeconds,
            discardedBacklog,
        });
    }
}

function toFiniteSeconds(timestampMilliseconds) {
    if (!Number.isFinite(timestampMilliseconds)) {
        throw new TypeError("timestampMilliseconds must be finite");
    }
    return timestampMilliseconds / 1000;
}
